File: scripts/validate-contracts.py
#!/usr/bin/env python3
"""Validate ores.otel.log schemas, fixtures, SDK manifests, and migration matrix."""
from __future__ import annotations

import json
import os
import re
import sys
from pathlib import Path
from typing import Any, Iterable

from jsonschema import Draft202012Validator, FormatChecker
from referencing import Registry, Resource

ROOT = Path(__file__).resolve().parents[1]
CONTRACTS = ROOT / "contracts"
CANONICAL = "ores-otel/ores.otel.log"
LEGACY = "ORESoftware/next-loggers.ts"
LANGUAGES = {"nodejs", "python", "go", "rust", "java", "dart", "gleam", "erlang", "elixir", "ruby", "wasm"}
REQUIRED_CAPABILITIES = {"record-schema", "api-manifest", "context-isolation", "otel-explicit", "no-monkey-patching", "lifecycle"}
FORBIDDEN = (
    ("automatic instrumentation", re.compile(r"registerInstrumentations\s*\(")),
    ("global tracer provider", re.compile(r"setGlobalTracerProvider\s*\(")),
    ("global meter provider", re.compile(r"setGlobalMeterProvider\s*\(")),
    ("global logger provider", re.compile(r"setGlobalLoggerProvider\s*\(")),
    ("require-in-the-middle", re.compile(r"require-in-the-middle", re.I)),
    ("shimmer", re.compile(r"(?:from|require\s*\()\s*['\"]shimmer['\"]", re.I)),
    ("Node module loader replacement", re.compile(r"Module\s*\.\s*_load\s*=(?!=|>)")),
    ("prototype replacement", re.compile(r"\.\s*prototype\s*\.[A-Za-z_$][\w$]*\s*=(?!=|>)")),
    ("console replacement", re.compile(r"console\s*\.\s*(?:log|debug|info|warn|error)\s*=(?!=|>)")),
    ("global fetch replacement", re.compile(r"globalThis\s*\.\s*fetch\s*=(?!=|>)")),
)

class Failure(RuntimeError):
    pass


def load(path: Path) -> Any:
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except FileNotFoundError as exc:
        raise Failure(f"missing required file: {path.relative_to(ROOT)}") from exc
    except json.JSONDecodeError as exc:
        raise Failure(f"invalid JSON in {path.relative_to(ROOT)}:{exc.lineno}:{exc.colno}: {exc.msg}") from exc


def schema_documents() -> dict[str, tuple[Path, dict[str, Any]]]:
    paths = [CONTRACTS / "log-record.schema.json", *sorted((CONTRACTS / "schemas").glob("*.json"))]
    documents: dict[str, tuple[Path, dict[str, Any]]] = {}
    for path in paths:
        document = load(path)
        Draft202012Validator.check_schema(document)
        identifier = document.get("$id")
        if not isinstance(identifier, str) or "ores-otel/ores.otel.log" not in identifier:
            raise Failure(f"{path.relative_to(ROOT)} must use a canonical ores-otel $id")
        if identifier in documents:
            raise Failure(f"duplicate schema $id: {identifier}")
        documents[identifier] = (path, document)
    return documents


def registry_for(documents: dict[str, tuple[Path, dict[str, Any]]]) -> Registry[Any]:
    registry: Registry[Any] = Registry()
    for identifier, (_, document) in documents.items():
        registry = registry.with_resource(identifier, Resource.from_contents(document))
    return registry


def validator(path: Path, registry: Registry[Any]) -> Draft202012Validator:
    return Draft202012Validator(load(path), registry=registry, format_checker=FormatChecker())


def keyword_set(error: Any) -> set[str]:
    values = {str(error.validator)}
    for child in error.context:
        values.update(keyword_set(child))
    return values


def first_error(errors: Iterable[Any]) -> str:
    error = next(iter(errors))
    location = "/".join(map(str, error.absolute_path)) or "<root>"
    return f"{location}: {error.message}"


def validate_fixtures(registry: Registry[Any]) -> int:
    manifest_path = CONTRACTS / "fixtures" / "manifest.json"
    manifest = load(manifest_path)
    errors = sorted(validator(CONTRACTS / "schemas" / "fixture-manifest.schema.json", registry).iter_errors(manifest), key=lambda e: list(e.absolute_path))
    if errors:
        raise Failure(f"invalid fixture manifest: {first_error(errors)}")
    seen: set[str] = set()
    for case in manifest["cases"]:
        fixture_path = ROOT / case["path"]
        if case["path"] in seen:
            raise Failure(f"duplicate fixture case: {case['path']}")
        seen.add(case["path"])
        errors = sorted(validator(ROOT / case["schema"], registry).iter_errors(load(fixture_path)), key=lambda e: (list(e.absolute_path), e.message))
        if case["valid"] and errors:
            raise Failure(f"expected valid fixture {case['path']}: {first_error(errors)}")
        if not case["valid"]:
            if not errors:
                raise Failure(f"expected invalid fixture to fail: {case['path']}")
            observed: set[str] = set()
            for error in errors:
                observed.update(keyword_set(error))
            if case["expectedKeyword"] not in observed:
                raise Failure(f"{case['path']} failed for {sorted(observed)}, not {case['expectedKeyword']!r}")
    return len(manifest["cases"])


def validate_manifests(registry: Registry[Any]) -> tuple[int, list[dict[str, Any]]]:
    index_path = CONTRACTS / "sdk-manifests.json"
    index = load(index_path)
    errors = list(validator(CONTRACTS / "schemas" / "sdk-manifest-set.schema.json", registry).iter_errors(index))
    if errors:
        raise Failure(f"invalid SDK manifest index: {first_error(errors)}")
    manifest_validator = validator(CONTRACTS / "schemas" / "sdk-manifest.schema.json", registry)
    manifests: list[dict[str, Any]] = []
    seen: set[str] = set()
    for relative in index["manifests"]:
        manifest = load(ROOT / relative)
        errors = sorted(manifest_validator.iter_errors(manifest), key=lambda e: (list(e.absolute_path), e.message))
        if errors:
            raise Failure(f"invalid SDK manifest {relative}: {first_error(errors)}")
        language = manifest["language"]
        if language in seen:
            raise Failure(f"duplicate SDK manifest language: {language}")
        seen.add(language)
        otel = manifest["telemetry"]["opentelemetry"]
        if any((otel["automaticInstrumentation"], otel["monkeyPatching"], otel["ownsProvider"])):
            raise Failure(f"{language} enables forbidden OTEL behavior")
        if not otel["explicit"] or not otel["wrapsLoggerCalls"]:
            raise Failure(f"{language} OTEL must remain explicit and downstream of logger calls")
        promotion = manifest["promotion"]
        if promotion["ready"]:
            required = {
                "implemented context": manifest["context"]["status"] == "implemented",
                "schema validation": manifest["conformance"]["validatesJsonSchema"],
                "context isolation tests": manifest["conformance"]["testsContextIsolation"],
                "anti-patching tests": manifest["conformance"]["testsNoMonkeyPatching"],
                "sampled-out correlation": otel["sampledOutCorrelation"],
                "recording-only span events": otel["recordingSpanEventsOnly"],
            }
            missing = [name for name, value in required.items() if not value]
            if missing:
                raise Failure(f"{language} is promotion-ready without {', '.join(missing)}")
        elif not promotion["blockers"]:
            raise Failure(f"{language} is not ready but has no explicit blockers")
        if os.environ.get("ORES_OTEL_SKIP_SOURCE_CHECK") != "1":
            if not (ROOT / manifest["sourceRoot"]).is_dir():
                raise Failure(f"{language} source root does not exist: {manifest['sourceRoot']}")
            for source in manifest["conformance"]["sourceFiles"]:
                source_path = ROOT / source
                if not source_path.is_file():
                    raise Failure(f"{language} source file does not exist: {source}")
                text = source_path.read_text(encoding="utf-8")
                for description, pattern in FORBIDDEN:
                    if pattern.search(text):
                        raise Failure(f"{description} is forbidden in {source}")
        manifests.append(manifest)
    if seen != LANGUAGES:
        raise Failure(f"SDK manifest set drifted; missing={sorted(LANGUAGES-seen)}, extra={sorted(seen-LANGUAGES)}")
    return len(manifests), manifests


def validate_repository_identity(manifests: list[dict[str, Any]]) -> None:
    """Keep canonical and legacy Go module identities explicit during migration."""
    go_manifest = next((manifest for manifest in manifests if manifest["language"] == "go"), None)
    if go_manifest is None:
        raise Failure("Go SDK manifest is missing")
    repository_context = (
        os.environ.get("ORES_OTEL_SOURCE_REPOSITORY")
        or os.environ.get("EXPECTED_REPOSITORY")
        or os.environ.get("GITHUB_REPOSITORY")
        or CANONICAL
    )
    package_names = go_manifest["package"]
    expected = (
        package_names["legacyName"]
        if repository_context == LEGACY
        else package_names["canonicalName"]
    )
    go_mod = ROOT / "sdk" / "go" / "go.mod"
    first_line = go_mod.read_text(encoding="utf-8").splitlines()[0]
    if first_line != f"module {expected}":
        raise Failure(
            "Go module identity does not match the source repository context: "
            f"repository={repository_context!r}, expected={expected!r}, observed={first_line!r}"
        )

def validate_migration(registry: Registry[Any], manifests: list[dict[str, Any]]) -> int:
    aliases = load(CONTRACTS / "migration" / "repository-aliases.json")
    matrix = load(CONTRACTS / "migration" / "test-repository-matrix.json")
    for document, schema in (
        (aliases, CONTRACTS / "schemas" / "repository-aliases.schema.json"),
        (matrix, CONTRACTS / "schemas" / "test-repository-matrix.schema.json"),
    ):
        errors = list(validator(schema, registry).iter_errors(document))
        if errors:
            raise Failure(f"invalid {schema.name}: {first_error(errors)}")
    repositories = matrix["repositories"]
    names = [repo["name"] for repo in repositories]
    if len(names) != len(set(names)):
        raise Failure("test repository names must be unique")
    languages = {repo["language"] for repo in repositories}
    requirements = aliases["promotionRequirements"]
    if len(repositories) < requirements["minimumRepositories"] or len(languages) < requirements["minimumLanguages"]:
        raise Failure("test matrix does not meet repository/language minimums")
    manifest_languages = {manifest["language"] for manifest in manifests}
    for repo in repositories:
        if set(repo["sources"]) != {CANONICAL, LEGACY}:
            raise Failure(f"{repo['name']} must test canonical and legacy sources exactly")
        if not REQUIRED_CAPABILITIES.issubset(repo["capabilities"]):
            raise Failure(f"{repo['name']} is missing required capabilities")
        if repo["language"] not in manifest_languages | {"browser"}:
            raise Failure(f"{repo['name']} has no SDK manifest for {repo['language']}")
        if not repo["name"].startswith("ores-otel-log-"):
            raise Failure(f"invalid test repository prefix: {repo['name']}")
    return len(repositories)


def main() -> int:
    try:
        documents = schema_documents()
        registry = registry_for(documents)
        fixtures = validate_fixtures(registry)
        manifest_count, manifests = validate_manifests(registry)
        validate_repository_identity(manifests)
        repositories = validate_migration(registry, manifests)
    except Failure as error:
        print(f"contract validation failed: {error}", file=sys.stderr)
        return 1
    print(f"contract validation passed: {len(documents)} schemas, {fixtures} fixtures, {manifest_count} SDK manifests, {repositories} test repositories")
    return 0

if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/validate-contracts.mjs
#!/usr/bin/env node

import { access, readFile } from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
const contractPath = path.join(root, 'contracts', 'logger-api.json');
const contractSchemaPath = path.join(root, 'contracts', 'logger-api.schema.json');
const matrixPath = path.join(root, 'contracts', 'test-org-matrix.json');
const matrixSchemaPath = path.join(root, 'contracts', 'test-org-matrix.schema.json');
const recordSchemaPath = path.join(root, 'contracts', 'log-record.schema.json');

const expectedOperations = {
  logger: ['create', 'trace', 'debug', 'info', 'warn', 'error', 'fatal', 'flush', 'flush_on_exit', 'close'],
  event: ['field', 'fields', 'user', 'users', 'trace', 'traces', 'routine', 'tag', 'tags', 'context', 'meta', 'error', 'send'],
  transport: ['write', 'flush', 'flush_on_exit', 'close'],
  context: ['current', 'run', 'bind', 'set', 'get', 'clear'],
  otel: ['extract', 'inject', 'correlate', 'create_transport', 'preserve_provider_ownership', 'no_global_patching'],
};

const requiredCapabilities = [
  'structured-records',
  'otel-explicit',
  'no-global-patching',
  'transport-lifecycle',
  'deterministic-conformance',
];

const credentialPattern = new RegExp([
  'gh[pousr]_[A-Za-z0-9]{20,}',
  'lin_' + 'api_[A-Za-z0-9]+',
  'cfa' + 't_[A-Za-z0-9]+',
  'BEGIN [A-Z ]*PRIVATE KEY',
  'CHAT_' + 'BRIDGE_TOKEN',
].join('|'), 'u');

function fail(message) {
  throw new Error(message);
}

function assert(condition, message) {
  if (!condition) fail(message);
}

function deepEqual(left, right) {
  return JSON.stringify(left) === JSON.stringify(right);
}

function valueType(value) {
  if (value === null) return 'null';
  if (Array.isArray(value)) return 'array';
  if (Number.isInteger(value)) return 'integer';
  return typeof value;
}

function resolveRef(rootSchema, ref) {
  assert(ref.startsWith('#/'), `only local JSON Schema refs are supported: ${ref}`);
  let current = rootSchema;
  for (const rawPart of ref.slice(2).split('/')) {
    const part = rawPart.replaceAll('~1', '/').replaceAll('~0', '~');
    assert(current && Object.hasOwn(current, part), `unresolved JSON Schema ref: ${ref}`);
    current = current[part];
  }
  return current;
}

function schemaErrors(value, schema, rootSchema, location = '$') {
  if (schema.$ref) {
    return schemaErrors(value, resolveRef(rootSchema, schema.$ref), rootSchema, location);
  }

  const errors = [];

  if (schema.oneOf) {
    const matches = schema.oneOf.filter(candidate =>
      schemaErrors(value, candidate, rootSchema, location).length === 0
    ).length;
    if (matches !== 1) errors.push(`${location}: expected exactly one oneOf branch, matched ${matches}`);
    return errors;
  }

  if (Object.hasOwn(schema, 'const') && !deepEqual(value, schema.const)) {
    errors.push(`${location}: expected const ${JSON.stringify(schema.const)}`);
  }

  if (schema.enum && !schema.enum.some(candidate => deepEqual(value, candidate))) {
    errors.push(`${location}: expected one of ${JSON.stringify(schema.enum)}`);
  }

  if (schema.type) {
    const actual = valueType(value);
    const valid = schema.type === 'number'
      ? actual === 'integer' || actual === 'number'
      : actual === schema.type;
    if (!valid) {
      errors.push(`${location}: expected ${schema.type}, received ${actual}`);
      return errors;
    }
  }

  if (typeof value === 'string') {
    if (schema.minLength !== undefined && value.length < schema.minLength) {
      errors.push(`${location}: expected minLength ${schema.minLength}`);
    }
    if (schema.pattern && !new RegExp(schema.pattern, 'u').test(value)) {
      errors.push(`${location}: did not match ${schema.pattern}`);
    }
  }

  if (typeof value === 'number' && schema.minimum !== undefined && value < schema.minimum) {
    errors.push(`${location}: expected minimum ${schema.minimum}`);
  }

  if (Array.isArray(value)) {
    if (schema.minItems !== undefined && value.length < schema.minItems) {
      errors.push(`${location}: expected at least ${schema.minItems} items`);
    }
    if (schema.maxItems !== undefined && value.length > schema.maxItems) {
      errors.push(`${location}: expected at most ${schema.maxItems} items`);
    }
    if (schema.uniqueItems) {
      const serialized = value.map(item => JSON.stringify(item));
      if (new Set(serialized).size !== serialized.length) {
        errors.push(`${location}: expected uniqueItems`);
      }
    }
    if (schema.items) {
      value.forEach((item, index) => {
        errors.push(...schemaErrors(item, schema.items, rootSchema, `${location}[${index}]`));
      });
    }
    if (schema.contains) {
      const contains = value.some(item => schemaErrors(item, schema.contains, rootSchema, location).length === 0);
      if (!contains) errors.push(`${location}: no item matched contains`);
    }
  }

  if (value && typeof value === 'object' && !Array.isArray(value)) {
    for (const key of schema.required ?? []) {
      if (!Object.hasOwn(value, key)) errors.push(`${location}: missing required property ${key}`);
    }
    for (const [key, child] of Object.entries(value)) {
      if (schema.properties?.[key]) {
        errors.push(...schemaErrors(child, schema.properties[key], rootSchema, `${location}.${key}`));
      } else if (schema.additionalProperties === false) {
        errors.push(`${location}: unexpected property ${key}`);
      } else if (schema.additionalProperties && typeof schema.additionalProperties === 'object') {
        errors.push(...schemaErrors(child, schema.additionalProperties, rootSchema, `${location}.${key}`));
      }
    }
  }

  return errors;
}

async function loadJson(file) {
  const raw = await readFile(file, 'utf8');
  assert(!credentialPattern.test(raw), `${path.relative(root, file)} contains credential-shaped content`);
  try {
    return { value: JSON.parse(raw) };
  } catch (error) {
    fail(`${path.relative(root, file)} is not valid JSON: ${error.message}`);
  }
}

function validateWithSchema(name, value, schema) {
  const errors = schemaErrors(value, schema, schema);
  if (errors.length > 0) {
    fail(`${name} failed JSON Schema validation:\n${errors.slice(0, 50).map(error => `- ${error}`).join('\n')}`);
  }
}

function assertExactOperationSet(contract) {
  for (const [category, expected] of Object.entries(expectedOperations)) {
    const actual = contract.operations[category].map(operation => operation.id);
    assert(
      deepEqual(actual, expected),
      `${category} operation drift: expected ${expected.join(', ')}, received ${actual.join(', ')}`
    );
    assert(new Set(actual).size === actual.length, `${category} contains duplicate operation IDs`);
  }
}

async function validateSdkBindings(contract) {
  assert(contract.sdkBindings.length >= 11, 'at least 11 SDK bindings are required');
  const ids = contract.sdkBindings.map(binding => binding.id);
  assert(new Set(ids).size === ids.length, 'SDK binding IDs must be unique');

  for (const binding of contract.sdkBindings) {
    await access(path.join(root, binding.path));
    assert(binding.operationSet === contract.operationSet, `${binding.id} uses the wrong operation set`);
    for (const capability of requiredCapabilities) {
      assert(binding.capabilities.includes(capability), `${binding.id} is missing ${capability}`);
    }
    assert(
      binding.capabilities.includes('context-local-storage') ||
        binding.capabilities.includes('explicit-context-passing'),
      `${binding.id} must declare context-local-storage or explicit-context-passing`
    );
  }
}

function validateTracks(matrix) {
  const tracks = new Map(matrix.tracks.map(track => [track.id, track]));
  assert(tracks.size === 2, 'matrix must define exactly legacy and canonical tracks');
  const legacy = tracks.get('legacy');
  const canonical = tracks.get('canonical');
  assert(legacy?.repository === 'ORESoftware/next-loggers.ts', 'legacy repository drifted');
  assert(/^[0-9a-f]{40}$/u.test(legacy?.ref ?? ''), 'legacy track requires an exact 40-character ref');
  assert(legacy.status === 'ready', 'legacy track must be ready');
  assert(canonical?.repository === 'ores-otel/ores.otel.log', 'canonical repository drifted');
  if (canonical.status === 'ready') {
    assert(/^[0-9a-f]{40}$/u.test(canonical.ref ?? ''), 'ready canonical track requires an exact ref');
  } else {
    assert(canonical.status === 'blocked', 'canonical status must be ready or blocked');
    assert(canonical.ref === null, 'blocked canonical track must not claim a ref');
    assert(Boolean(canonical.blockedBy), 'blocked canonical track requires blockedBy');
  }
}

function validateMatrix(contract, matrix) {
  assert(matrix.targetOrganization === 'ores-otel-test', 'test-org writes are allowlisted to ores-otel-test');
  assert(matrix.policy.productionWritesAllowed === false, 'test matrix must prohibit production writes');
  assert(matrix.repositories.length >= Math.max(10, matrix.policy.minimumRepositories), 'too few test repositories');

  const languages = new Set(matrix.repositories.map(repository => repository.language));
  assert(languages.size >= Math.max(7, matrix.policy.minimumLanguages), 'too few test languages');
  assert(
    languages.size === contract.sdkBindings.length,
    `matrix must cover every SDK binding (${contract.sdkBindings.length}), found ${languages.size}`
  );

  const names = matrix.repositories.map(repository => repository.name);
  assert(new Set(names).size === names.length, 'test repository names must be unique');

  const bindings = new Map(contract.sdkBindings.map(binding => [binding.id, binding]));
  for (const binding of contract.sdkBindings) {
    const entries = matrix.repositories.filter(repository => repository.language === binding.id);
    assert(entries.length === 2, `${binding.id} requires exactly legacy and canonical consumers`);
    assert(
      deepEqual(entries.map(entry => entry.track).sort(), ['canonical', 'legacy']),
      `${binding.id} consumer pair is incomplete`
    );
    for (const entry of entries) {
      assert(entry.visibility === 'private', `${entry.name} must remain private`);
      assert(entry.sourcePath === binding.path, `${entry.name} sourcePath drifted`);
      assert(entry.package === binding.package, `${entry.name} package drifted`);
      assert(entry.testCommand === binding.testCommand, `${entry.name} testCommand drifted`);
      assert(entry.requiredChecks.includes('contract'), `${entry.name} must require contract`);
      assert(entry.requiredChecks.includes('consumer'), `${entry.name} must require consumer`);
    }
  }

  for (const entry of matrix.repositories) {
    assert(bindings.has(entry.language), `${entry.name} refers to unknown SDK ${entry.language}`);
  }

  validateTracks(matrix);
}

async function main() {
  const [
    { value: contract },
    { value: contractSchema },
    { value: matrix },
    { value: matrixSchema },
    { value: recordSchema },
  ] = await Promise.all([
    loadJson(contractPath),
    loadJson(contractSchemaPath),
    loadJson(matrixPath),
    loadJson(matrixSchemaPath),
    loadJson(recordSchemaPath),
  ]);

  assert(contractSchema.$schema === 'https://json-schema.org/draft/2020-12/schema', 'API schema draft drifted');
  assert(matrixSchema.$schema === 'https://json-schema.org/draft/2020-12/schema', 'matrix schema draft drifted');
  validateWithSchema('logger-api.json', contract, contractSchema);
  validateWithSchema('test-org-matrix.json', matrix, matrixSchema);

  assert(recordSchema.properties?.schema?.const === contract.wireSchema, 'wire schema and API contract disagree');
  assertExactOperationSet(contract);
  await validateSdkBindings(contract);
  validateMatrix(contract, matrix);

  process.stdout.write(
    `Validated ${contract.sdkBindings.length} SDK bindings, ` +
      `${Object.values(contract.operations).flat().length} canonical operations, ` +
      `${matrix.repositories.length} test repositories, and ` +
      `${new Set(matrix.repositories.map(repository => repository.language)).size} languages.\n`
  );
}

main().catch(error => {
  process.stderr.write(`${error.stack ?? error.message}\n`);
  process.exitCode = 1;
});
